diff --git a/libraries/liblmdb/mdb.c b/libraries/liblmdb/mdb.c
--- a/libraries/liblmdb/mdb.c
+++ b/libraries/liblmdb/mdb.c
@@ -6526,6 +6526,8 @@
 				MDB_GET_KEY(leaf, key);
 				if (data) {
 					if (F_ISSET(leaf->mn_flags, F_DUPDATA)) {
+						if (!mc->mc_xcursor)
+							return MDB_CORRUPTED;
 						rc = mdb_cursor_get(&mc->mc_xcursor->mx_cursor, data, NULL, MDB_GET_CURRENT);
 					} else {
 						rc = mdb_node_read(mc, leaf, data);
@@ -7301,6 +7303,8 @@
 	leaf = NODEPTR(mp, mc->mc_ki[mc->mc_top]);

 	if (F_ISSET(leaf->mn_flags, F_DUPDATA)) {
+		if (!mc->mc_xcursor)
+			return MDB_CORRUPTED;
 		if (flags & MDB_NODUPDATA) {
 			/* mdb_cursor_del0() will subtract the final entry */
 			mc->mc_db->md_entries -= mc->mc_xcursor->mx_db.md_entries - 1;
